# Amazon Product Intelligence Assistant

## Task – Data Engineering

### Step 1.1 – Dataset Exploration & Profiling

In [ ]:
# Required Imports
import pandas as pd
import numpy as np

from pathlib import Path

# To see full data
pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", 120)
pd.set_option("display.width", 200)

In [14]:
DATA_DIR = Path("/home/mansoor_ali/amazon-product-intelligence/data/raw")

products_df = pd.read_csv(DATA_DIR / "products.csv")
reviews_df = pd.read_csv(DATA_DIR / "reviews.csv")

* **Products Dataset Overview**

In [ ]:
# Overview
products_df.head()

,s.no,about_item,asin,availability,best_sellers_rank,brand_name,brand_page_url,breadcrumbs,customer_review_summary,default_variant/0,default_variant/1,default_variant/2,delivery_date,fastest_delivery_date,list_price,manufacturer,model_number,price_value,product_description,product_url,rating_count,rating_distribution/1star,rating_distribution/2star,rating_distribution/3star,rating_distribution/4star,rating_distribution/5star,rating_stars,recent_purchases,scrape_time,seller_name,seller_page_url,title,all_images,rank_1
0,0,"Premium Comfort: Crafted from a high-quality cotton blend, these mens polo shirts are soft, lightweight, and breatha...",B0B59BJG6Y,In Stock,"#56,836 in Clothing, Shoes & Jewelry (See Top 100 in Clothing, Shoes & Jewelry) #85 in Men's Activewear Polos",MLYENX Store,https://www.amazon.com/stores/MLYENX/page/1FCDBC12-B80F-4084-A727-D6DE911CC4EB?ref_=ast_bln&store_ref=bl_ast_dp_bran...,"Clothing, Shoes & Jewelry › Men › Clothing › Active › Active Shirts & Tees › Polos","Customers find the shirts comfortable and well-fitting. They appreciate the soft, smooth material that feels very sm...",size:Large,"color:5 Pack Black, Dark Grey, Light Blue, Milddle Blue, Light Grey",NaN,"Saturday, March 15","Tomorrow, March 11",List Price: $53.99,NaN,NaN,39.9926,NaN,https://www.amazon.com/dp/B0B59BJG6Y,"1,654 ratings",2%,1%,7%,15%,75%,4.6 out of 5 stars,50+ bought,03-10-2025 21:42,Greenfive,https://www.amazon.com/gp/help/seller/at-a-glance.html/ref=dp_merchant_link?ie=UTF8&seller=A5ETY7BZYGD5S&asin=B0B59B...,4/5 Pack Mens Polo Shirts Short Sleeve Quick Dry Moisture Wicking Casual Golf T Shirts for Men,"['https://m.media-amazon.com/images/I/41yUF65PmbL.jpg', 'https://m.media-amazon.com/images/I/41jMweQcgzL.jpg', 'http...",85.0
1,1,"Material: Men's polo shirt is made of soft polyester fabric, moisture wicking, stain-resistant, durable work shirts,...",B0DLGB4RYH,In Stock,"#50,261 in Clothing, Shoes & Jewelry (See Top 100 in Clothing, Shoes & Jewelry) #199 in Men's Polo Shirts",COOFANDY Store,https://www.amazon.com/stores/COOFANDY/page/23614F3D-BBD9-4564-8B56-4436A5CD5A93?ref_=ast_bln&store_ref=bl_ast_dp_br...,"Clothing, Shoes & Jewelry › Men › Clothing › Shirts › Polos",NaN,size:X-Large,color:Light Grey,NaN,"Saturday, March 15","Tomorrow, March 11",Typical price: $24.99,NaN,NaN,19.9920,NaN,https://www.amazon.com/dp/B0DLGB4RYH,20 ratings,0%,6%,11%,17%,66%,4.4 out of 5 stars,NaN,03-10-2025 21:42,COOFANDY,https://www.amazon.com/gp/help/seller/at-a-glance.html/ref=dp_merchant_link?ie=UTF8&seller=AW10J8VB8Z74G&asin=B0DLGB...,COOFANDY Men's Polo Shirts Short Sleeve Moisture Wicking Golf Shirt Fashion Casual Collared T-Shirt,"['https://m.media-amazon.com/images/I/31a3mSs3paL.jpg', 'https://m.media-amazon.com/images/I/41jqhQ7vMLL.jpg', 'http...",199.0
2,2,"PERFORMANCE:These men polo shirts are soft,lightweight and moisture wicking,keep you cool and breathableMULTIPACK:3 ...",B0DRXF62JH,In Stock,"#69,641 in Clothing, Shoes & Jewelry (See Top 100 in Clothing, Shoes & Jewelry) #111 in Men's Activewear Polos",ZITY Store,https://www.amazon.com/stores/ZITY/page/F582B6FA-FA45-4D47-9B10-7296650A2718?ref_=ast_bln&store_ref=bl_ast_dp_brandL...,"Clothing, Shoes & Jewelry › Men › Clothing › Active › Active Shirts & Tees › Polos",NaN,size:X-Large,color:Bblack Grey Navy,NaN,"Saturday, March 15","Wednesday, March 12",List Price: $59.99,NaN,NaN,29.9950,NaN,https://www.amazon.com/dp/B0DRXF62JH,5 ratings,0%,0%,0%,37%,63%,4.6 out of 5 stars,NaN,03-10-2025 21:42,ZITY®,https://www.amazon.com/gp/help/seller/at-a-glance.html/ref=dp_merchant_link?ie=UTF8&seller=A26KN7P3I0WWYV&asin=B0DRX...,ZITY 3 Pack Men Polo Shirts Short Sleeve with Pocket Sport Wicking Shirts for Men Casual Athletic Collared T-Shirts,"['https://m.media-amazon.com/images/I/41J5q-Ua0KL.jpg', 'https://m.media-amazon.com/images/I/410wWdbkEzL.jpg', 'http...",111.0
3,3,"【Material】: These golf shirts for men are made of quality polyester and spandx fabric, providing you with lightweigh...",B0DK5FZ

In [ ]:
# Shape of the data
products_df.shape

(728, 34)

In [ ]:
# Info
products_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 728 entries, 0 to 727
Data columns (total 34 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   s.no                       728 non-null    int64  
 1   about_item                 728 non-null    str    
 2   asin                       728 non-null    str    
 3   availability               715 non-null    str    
 4   best_sellers_rank          558 non-null    str    
 5   brand_name                 728 non-null    str    
 6   brand_page_url             649 non-null    str    
 7   breadcrumbs                715 non-null    str    
 8   customer_review_summary    634 non-null    str    
 9   default_variant/0          696 non-null    str    
 10  default_variant/1          658 non-null    str    
 11  default_variant/2          1 non-null      str    
 12  delivery_date              702 non-null    str    
 13  fastest_delivery_date      667 non-null    str    
 14  list_

In [ ]:
# Statictics
products_df.describe(include="all")

,s.no,about_item,asin,availability,best_sellers_rank,brand_name,brand_page_url,breadcrumbs,customer_review_summary,default_variant/0,default_variant/1,default_variant/2,delivery_date,fastest_delivery_date,list_price,manufacturer,model_number,price_value,product_description,product_url,rating_count,rating_distribution/1star,rating_distribution/2star,rating_distribution/3star,rating_distribution/4star,rating_distribution/5star,rating_stars,recent_purchases,scrape_time,seller_name,seller_page_url,title,all_images,rank_1
count,728.000000,728,728,715,558,728,649,715,634,696,658,1,702,667,401,266,199,707.000000,271,728,719,728,728,728,728,728,719,569,728,707,426,728,728,540.000000
unique,NaN,697,728,11,539,285,237,140,616,118,413,1,51,33,205,131,191,NaN,238,728,645,34,19,24,33,63,24,19,120,267,426,704,714,NaN
top,NaN,"Textured fabric that's soft, light & breathable|Material wicks sweat & dries really fast|Anti-odor technology preven...",B0B59BJG6Y,In Stock,"#67 in Clothing, Shoes & Jewelry (See Top 100 in Clothing, Shoes & Jewelry) #1 in Men's Activewear Polos",Hanes Store,https://www.amazon.com/stores/Hanes/page/F83022EC-A709-4B2F-BCB7-7BBFA1896AB0?ref_=ast_bln&store_ref=bl_ast_dp_brand...,"Clothing, Shoes & Jewelry › Men › Clothing › Jeans","Customers find the shirt has a good fit and style. They appreciate the flattering shoulders, collar, and extra lengt...",size:Large,color:Black,color:Team Red/Black/Team Red Viper,"Saturday, March 15","Tomorrow, March 11",List Price: $39.99,Amazon Essentials,1290140,NaN,"We listen to customer feedback and fine-tune every detail to ensure our clothes are more comfortable, higher quality...",https://www.amazon.com/dp/B0B59BJG6Y,20 ratings,3%,2%,6%,15%,75%,4.5 out of 5 stars,100+ bought,03-10-2025 22:38,Amazon.com,https://www.amazon.com/gp/help/seller/at-a-glance.html/ref=dp_merchant_link?ie=UTF8&seller=A5ETY7BZYGD5S&asin=B0B59B...,Under Armour Men's Tech Golf Polo,"['https://m.media-amazon.com/images/I/21+XaygRfPL.jpg', 'https://m.media-amazon.com/images/I/31fTk6PXieL.jpg', 'http...",NaN
freq,NaN,3,1,640,3,35,35,52,3,168,106,1,610,491,10,21,3,NaN,8,1,4,123,194,107,89,44,131,108,18,278,1,3,2,NaN
mean,423.361264,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,35.323948,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,174.622222
std,293.923401,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,26.422336,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,240.115097
min,0.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.457300,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000
25%,184.750000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,19.990000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,18.000000
50%,371.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,28.950000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,56.500000
75%,553.250000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,41.771600,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,220.250000


In [23]:
products_df.memory_usage(deep=True).sum() / 1024**2

np.float64(1.666569709777832)

## 📊 Dataset Exploration – Products Dataset (`products.csv`)

### Dataset Overview

The `products.csv` dataset contains product-level information scraped from Amazon. Each record represents a single product and includes product metadata, descriptive text, pricing information, ratings, seller information, and various scrape-specific attributes.

| Property | Value |
|----------|-------|
| Dataset | `products.csv` |
| Rows | **728** |
| Columns | **34** |
| Memory Usage | **~1.7 MB** |
| Data Types | 31 String, 2 Float, 1 Integer |
| Candidate Primary Key | `asin` |
| Duplicate ASINs | None (728 unique products) |

---

## Initial Observations

- The dataset is relatively **small**, making it suitable for an in-memory processing pipeline using **Pandas**.
- Every product has a unique **ASIN**, which will serve as the **primary key** for the project and the join key with the `reviews.csv` dataset.
- The dataset appears to be generated from **web scraping** rather than an official Amazon catalog, meaning several columns represent scrape-time information rather than intrinsic product attributes.
- The majority of columns are textual, which is beneficial for building a Retrieval-Augmented Generation (RAG) knowledge base.

---

# Schema Classification

To simplify downstream preprocessing and document construction, columns are classified based on their intended role within the RAG system.

---

## 1. Primary Identifier

| Column | Decision | Reason |
|--------|----------|--------|
| `asin` |  Keep | Primary key and join key with reviews |
| `product_url` | Keep (Metadata) | Useful for source attribution and UI links |
| `s.no` |  Drop | Internal row identifier with no business value |

---

## 2. Core Product Content (Document Body)

These columns contain the primary semantic information that should form the textual document embedded into the vector database.

| Column | Decision | Reason |
|--------|----------|--------|
| `title` |  Keep | Strong retrieval signal |
| `about_item` |  Keep | Rich product feature list |
| `product_description` |  Keep | Detailed product description (when available) |
| `brand_name` |  Keep | Important product attribute |

---

## 3. Product Metadata

These fields are useful during retrieval and response generation but should primarily be stored as metadata rather than embedded directly into the document text.

| Column | Decision | Reason |
|--------|----------|--------|
| `price_value` |  Keep | Product price |
| `rating_stars` |  Keep | Overall rating |
| `rating_count` |  Keep | Rating confidence indicator |
| `best_sellers_rank` |  Keep | Popularity indicator |
| `breadcrumbs` |  Keep | Product category hierarchy |
| `availability` |  Keep | Inventory status |

---

## 4. Review Summary

| Column | Decision | Reason |
|--------|----------|--------|
| `customer_review_summary` |  Tentatively Keep | Will be evaluated after analyzing `reviews.csv` |

Since a separate `reviews.csv` dataset exists, the usefulness of this generated summary will be evaluated later before making a final decision.

---

## 5. Product Variants

| Column | Decision | Reason |
|--------|----------|--------|
| `default_variant/0` |  Keep | Variant information (e.g., size) |
| `default_variant/1` |  Keep | Variant information (e.g., color) |
| `default_variant/2` |  Drop | Contains only one non-null value |

---

## 6. Manufacturer Information

| Column | Decision | Reason |
|--------|----------|--------|
| `manufacturer` |  Keep | Optional manufacturer metadata |
| `model_number` |  Keep | Useful for product identification and comparisons |

---

## 7. Seller Information

| Column | Decision | Reason |
|--------|----------|--------|
| `seller_name` |  Keep | Useful metadata |
| `seller_page_url` |  Drop | Not useful for retrieval |
| `brand_page_url` |  Drop | External navigation link |

---

## 8. Delivery Information

| Column | Decision | Reason |
|--------|----------|--------|
| `delivery_date` |  Drop | Time-dependent scrape information |
| `fastest_delivery_date` |  Drop | Time-dependent scrape information |

These values become stale quickly and do not contribute to semantic retrieval.

---

## 9. Rating Distribution

| Column | Decision | Reason |
|--------|----------|--------|
| `rating_distribution/1star` |  Keep | Customer sentiment metadata |
| `rating_distribution/2star` |  Keep | Customer sentiment metadata |
| `rating_distribution/3star` |  Keep | Customer sentiment metadata |
| `rating_distribution/4star` |  Keep | Customer sentiment metadata |
| `rating_distribution/5star` |  Keep | Customer sentiment metadata |

These fields may help answer questions regarding customer satisfaction and rating distribution.

---

## 10. Other Columns

| Column | Decision | Reason |
|--------|----------|--------|
| `list_price` |  Keep | Optional pricing metadata |
| `recent_purchases` |  Keep | Popularity indicator |
| `all_images` |  Drop | Image URLs are outside the scope of this text-based RAG system |
| `scrape_time` |  Drop | Operational metadata only |
| `rank_1` |  Investigate | Requires further understanding of its meaning |

---

# Missing Value Analysis

Several columns contain a significant number of missing values.

| Column | Non-Null Count | Missing (%) | Initial Decision |
|---------|---------------:|------------:|------------------|
| `product_description` | 271 | **62.8%** | Keep; fallback to `about_item` when missing |
| `manufacturer` | 266 | **63.5%** | Optional metadata |
| `model_number` | 199 | **72.7%** | Keep if available |
| `list_price` | 401 | **44.9%** | Optional metadata |
| `default_variant/2` | 1 | **99.9%** | Drop |

---

# Initial RAG Document Design

Based on the current analysis, each product document is expected to contain information similar to the following:

```text
Product:
Under Armour Men's Tech Golf Polo

Brand:
Under Armour

Category:
Clothing > Men > Shirts

Price:
$39.99

Average Rating:
4.5 / 5

About:
...

Description:
...

Customer Review Summary:
...
```

Structured attributes such as `asin`, `price_value`, `rating_count`, and `breadcrumbs` will likely be stored as **metadata** within the vector database rather than embedded into the document text.

---

# Preliminary Conclusions

- `products.csv` will serve as the **primary source of product knowledge** for the RAG system.
- The **ASIN** will be adopted as the canonical product identifier throughout the project.
- Product information should be divided into **document content** and **retrieval metadata** to improve retrieval quality and maintain a clean document structure.
- Final decisions regarding review summaries and document composition will be made after analyzing the `reviews.csv` dataset.
- The preprocessing pipeline should be designed to handle missing values gracefully, especially for optional descriptive fields such as `product_description`.